internal/spoa.go

```diff
@@ -229,3 +229,169 @@ func New(conf *config.Config) (*SPOA, error) {
 		defaultApplication: conf.DefaultApplication,
 	}, nil
 }
+
+func (s *SPOA) getApplication(appName string) (*application, error) {
+	var app *application
+
+	// Looking for app by name from message
+	if appName != "" {
+		app, exist := s.applications[appName]
+		if exist {
+			return app, nil
+		}
+	}
+
+	// Looking for app by default app name
+	app, exist := s.applications[s.defaultApplication]
+	if exist {
+		app.logger.Debug("application not found, using default", zap.Any("application", appName), zap.String("default", s.defaultApplication))
+		return app, nil
+	}
+
+	return nil, fmt.Errorf("application not found, application %s, default: %s", appName, s.defaultApplication)
+}
+
+func (s *SPOA) processRequest(spoeMsg spoe.Message) ([]spoe.Action, error) {
+	var (
+		err error
+		req *request
+		app *application
+		tx  types.Transaction
+	)
+
+	defer func() {
+		if tx == nil || app == nil {
+			return
+		}
+		if tx.IsInterrupted() {
+			tx.ProcessLogging()
+			if err := tx.Close(); err != nil {
+				app.logger.Error("failed to close transaction", zap.String("transaction_id", tx.ID()), zap.String("error", err.Error()))
+			}
+		} else {
+			if app.cfg.NoResponseCheck {
+				return
+			}
+			err := app.cache.SetWithExpire(tx.ID(), tx, time.Millisecond*time.Duration(app.cfg.TransactionTTLMilliseconds))
+			if err != nil {
+				app.logger.Error(fmt.Sprintf("failed to cache transaction: %s", err.Error()))
+			}
+		}
+	}()
+
+	req, err = NewRequest(spoeMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	app, err = s.getApplication(req.app)
```

Comment:
I would get the app from the spoeMsg and get the request right before we would use it.

Reply:
I don't get it. What problem do you try to solve?

From `spoeMsg` only application name can be obtained. Application != application name.

Also in request `request` and `respone` application name is already converted to string and 'validated'.

Reply:
Is [c490c9f](https://github.com/corazawaf/coraza-spoa/pull/68/commits/c490c9f2e499bd277a9005d76bb207cb591fc6fc) looks OK?